src/plugins/plugin-http.js

```diff
@@ -0,0 +1,114 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+var shimmer = require('shimmer');
+var url = require('url');
+var isString = require('is').string;
+var merge = require('lodash.merge');
+var httpAgent = require('_http_agent');
+
+function requestWrap(api, request) {
+  var labels = api.labels;
+  function setTraceHeader(parsedOptions, context) {
+    if (context) {
+      return merge(parsedOptions, {
+        headers: {
+          [TraceLabels.TRACE_CONTEXT_HEADER_NAME]: context
+        }
+      });
+    }
+    return parsedOptions;
+  }
+  function parseRequestOptions(requestOptions) {
+    return isString(requestOptions) ?
+      merge(url.parse(requestOptions), {headers: {}}) :
+      merge({headers: {}}, requestOptions);
+  }
+  function extractUrl(parsedOptions) {
+    var uri = parsedOptions;
+    var agent = parsedOptions._defaultAgent || httpAgent.globalAgent;
+    return isString(uri) ? uri :
+      (parsedOptions.protocol || agent.protocol) + '//' +
+      (parsedOptions.hostname || parsedOptions.host || 'localhost') +
+      ((isString(parsedOptions.port) ? (':' + parsedOptions.port) : '')) +
+      (parsedOptions.path || parseRequestOptions.pathName || '/');
+  }
+  function getSpanName(requestOptions) {
+    if (isString(options)) {
+      options = url.parse(requestOptions);
+    }
+    // c.f. _http_client.js ClientRequest constructor
+    return options.hostname || options.host || 'localhost';
+  }
+  function patchedHTTPRequest(requestOptions, callback, request) {
+    var parsedOptions = parseRequestOptions(requestOptions);
+    var uri = extractUrl(parsedOptions);
+    var requestLifecycleSpan = api.createChildSpan({name: 'http', url: uri})
+      .addLabel(api.labels.HTTP_METHOD_LABEL_KEY, parsedOptions.method)
+      .addLabel(api.labels.HTTP_URL_LABEL_KEY, uri);
+    parsedOptions = setTraceHeader(parsedOptions, transaction.getTraceContext());
+    var req = request.call(request, requestOptions, function (res) {
+      api.wrapEmitter(res);
+      var numBytes = 0;
+      res.on('data', function (chunk) {
+        numBytes += chunk.length;
+      });
+      res.on('end', function () {
+        requestLifecycleSpan
+          .addLabel(api.labels.HTTP_RESPONSE_SIZE_LABEL_KEY, numBytes);
+        requestLifecycleSpan
+          .addLabel(api.labels.HTTP_RESPONSE_CODE_LABEL_KEY, res.statusCode);
+        requestAnimationFrame.endSpan();
+      });
+      if (callback) {
+        return callback(res);
+      }
+    });
+    api.wrapEmitter(req);
+    req.on('error', function (e) {
+      var labels = {};
+      if (e) {
+        requestLifecycleSpan.addLabel(TraceLabels.ERROR_DETAILS_NAME, e.name);
+        requestLifecycleSpan
+          .addLabel(TraceLabels.ERROR_DETAILS_MESSAGE, e.message);
+      } else {
+        console.error('HTTP request error was null or undefined');
```

Comment:
I know this is in `hook-http.js` as well, but I don't think it's the responsibility of the trace agent to print an error if an `'error'` event is emitted without an argument. In addition, I'm not sure if `req` can even do this. Maybe we can omit this entirely?